File: app/Sources/LocalCaptionKit/TimeFormat.swift
import Foundation

/// Time/date formatting shared by the transcript file, per-line stamps, and the UI clock.
public enum TimeFormat {
    /// Seconds → `HH:MM:SS` (used for duration and the elapsed clock).
    public static func clock(_ seconds: Int) -> String {
        let s = max(0, seconds)
        return String(format: "%02d:%02d:%02d", s / 3600, (s % 3600) / 60, s % 60)
    }

    /// Milliseconds → `[HH:MM:SS]` per-line timestamp.
    public static func stamp(ms: Int) -> String { "[\(clock(ms / 1000))]" }

    /// `yyyy-MM-dd_HH-mm-ss` — filename-safe stamp from a start time.
    public static func fileStamp(_ date: Date) -> String { fmt("yyyy-MM-dd_HH-mm-ss", date) }

    /// `yyyy-MM-dd HH:mm:ss` — human header timestamp.
    public static func human(_ date: Date) -> String { fmt("yyyy-MM-dd HH:mm:ss", date) }

    /// `yyyy-MM-dd HH:mm` — short form for auto session names.
    public static func humanShort(_ date: Date) -> String { fmt("yyyy-MM-dd HH:mm", date) }

    /// ISO-8601 UTC — stored in the DB and in segment `created_at`.
    public static func iso(_ date: Date) -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f.string(from: date)
    }

    public static func parseISO(_ s: String) -> Date? {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f.date(from: s)
    }

    private static func fmt(_ pattern: String, _ date: Date) -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = pattern
        return f.string(from: date)
    }
}

File: windows/src/LocalCaption.Core/Transcripts/TimeFormat.cs
using System.Globalization;

namespace LocalCaption.Core.Transcripts;

/// <summary>
/// Time and date formatting shared by the transcript file, the per-line stamps and the UI
/// clock.
/// </summary>
/// <remarks>
/// Every format here is pinned to <see cref="CultureInfo.InvariantCulture"/> on purpose.
/// SPEC-WINDOWS.md §6: the ASUS's locale must not be able to reach into
/// <c>HH:mm:ss</c> or the ISO stamps, or transcripts stop being diffable against the macOS
/// output — which is how §17 verifies parity.
/// </remarks>
public static class TimeFormat
{
    /// <summary>Seconds → <c>HH:MM:SS</c> (duration and the elapsed clock).</summary>
    public static string Clock(int seconds)
    {
        var s = Math.Max(0, seconds);
        return string.Format(CultureInfo.InvariantCulture, "{0:00}:{1:00}:{2:00}",
                             s / 3600, s % 3600 / 60, s % 60);
    }

    /// <summary>Milliseconds → <c>[HH:MM:SS]</c> per-line timestamp.</summary>
    public static string Stamp(int ms) => $"[{Clock(ms / 1000)}]";

    /// <summary><c>yyyy-MM-dd_HH-mm-ss</c> — filename-safe stamp from a start time.</summary>
    public static string FileStamp(DateTimeOffset date) => Fmt("yyyy-MM-dd_HH-mm-ss", date);

    /// <summary><c>yyyy-MM-dd HH:mm:ss</c> — human header timestamp.</summary>
    public static string Human(DateTimeOffset date) => Fmt("yyyy-MM-dd HH:mm:ss", date);

    /// <summary><c>yyyy-MM-dd HH:mm</c> — short form for auto session names.</summary>
    public static string HumanShort(DateTimeOffset date) => Fmt("yyyy-MM-dd HH:mm", date);

    /// <summary>ISO-8601 UTC — stored in the DB and in each segment's <c>created_at</c>.</summary>
    public static string Iso(DateTimeOffset date) =>
        date.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture);

    public static DateTimeOffset? ParseIso(string s) =>
        DateTimeOffset.TryParse(s, CultureInfo.InvariantCulture,
                                DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                                out var parsed) ? parsed : null;

    private static string Fmt(string pattern, DateTimeOffset date) =>
        date.ToString(pattern, CultureInfo.InvariantCulture);
}
